compiler/rustc_middle/src/mir/mono.rs

```diff
@@ -274,13 +277,60 @@ pub struct MonoItemData {
     /// `GloballyShared` maps to `false` and `LocalCopy` maps to `true`.
     pub inlined: bool,
 
-    pub linkage: Linkage,
+    pub linkage_info: LinkageInfo,
     pub visibility: Visibility,
 
     /// A cached copy of the result of `MonoItem::size_estimate`.
     pub size_estimate: usize,
 }
 
+/// Stores how we know what linkage to use
+#[derive(Copy, Clone, PartialEq, Debug, TyEncodable, TyDecodable, HashStable)]
+pub enum LinkageInfo {
+    /// The linkage was specified explicitly (e.g. using #[linkage = "..."])
+    Explicit(Linkage),
+    /// Assume the symbol may be used from other CGUs, a safe default
+    ImplicitExternal,
+    /// We did not find any uses from other CGUs, so it's fine to make this internal
+    ImplicitInternal,
+}
+
+impl LinkageInfo {
+    pub const fn is_external(self) -> bool {
+        matches!(self.into_linkage(), Linkage::External)
+    }
+
+    pub const fn into_linkage(self) -> Linkage {
+        match self {
+            Self::Explicit(linkage) => linkage,
+            Self::ImplicitExternal => Linkage::External,
+            Self::ImplicitInternal => Linkage::Internal,
+        }
+    }
+
+    /// Linkage when the MonoItem is a naked function
+    ///
+    /// Naked functions are generated as a separate declaration (effectively an extern fn) and
+    /// definition (using global assembly). To link them together, some flavor of external linkage
+    /// must be used.
+    ///
+    /// This should be just an implementation detail of the backend, which is why this translation
+    /// is made at the final moment.
+    pub const fn into_naked_linkage(self) -> Linkage {
+        match self {
+            // promote Weak linkage to ExternalWeak
+            Self::Explicit(Linkage::WeakAny | Linkage::WeakODR) => Linkage::ExternalWeak,
```

Comment:
yeah, well, when we don't translate to `ExternalWeak`, the llvm validation fails in the `tests/run-make/naked-symbol-visibility` test:

```
Global is external, but doesn't have external or weak linkage!
ptr @_ZN12a_rust_dylib18naked_weak_linkage17h0826a2715adfb302E
```

Searching in the llvm source for that error eventually leads to

https://github.com/llvm/llvm-project/blob/39209724e66a20080bdebc609a051dfa0eb6b49f/llvm/include/llvm/IR/GlobalValue.h#L419

```cpp
  static bool isValidDeclarationLinkage(LinkageTypes Linkage) {
    return isExternalWeakLinkage(Linkage) || isExternalLinkage(Linkage);
  }
```

so, only `ExternalWeak` and `External` are valid linkages for global values. That could be an artificial LLVM limitation, but that is the status quo.

Reply:
We could of course generate an error when `linkage = "weak"` is used on a naked function, and suggest `extern_weak` instead? But I'll admit I don't fully understand the nuances of the linkage attribute (e.g. is using `extern_weak` a valid suggestion to make in that scenario; it'll compile then, but does it have the behavior you'd want?)

Reply:
I think the solution for that would be to skip `predefine_fn()` when a mono item is a naked function. `predefine_fn()` is supposed to be called in the codegen unit that will define the function using regular LLVM IR. If you call an extern function (as a naked function will effectively become with this PR), `get_fn` will declare the function as `Linkage::External` already to handle it as import rather than definition.